Add doc for Deploy API to Choreo Connect with Virtual Hosts (#3841)
wso2/docs-apim: en/docs/deploy-and-publish/deploy-on-gateway/choreo-connect/deploy-api/deploy-api-to-choreo-connect-with-virtual-hosts.md

# Deploy API to Choreo Connect with Virtual Hosts 

> There are two ways to add an API to the Choreo Connect.

Reviewer: ```suggestion
There are two ways to add an API to Choreo Connect.
```
Can we also add why we want to do this so that new users get a full picture when they land on this page?

Author: fixed with https://github.com/wso2/docs-apim/commit/b9900b303fe9c36a8933801e2dd3f8b605cab370 and add a reference to the overview page if the user wants more info.

1. [Via API Manager](#via-api-manager)
2. [Via the API Controller (apictl)](#via-the-api-controller-apictl)

!!! info
    **Before you begin**

    This guide assumes that you already have a Choreo Connect instance that is up and running.

>     If not, checkout the [Quick Start Guide]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/getting-started/quick-start-guide/quick-start-guide-with-apim)

Reviewer: ```suggestion
    If not, refer to the [Quick Start Guide]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/getting-started/quick-start-guide/quick-start-guide-with-apim)
```

Author: fixed with https://github.com/wso2/docs-apim/commit/b9900b303fe9c36a8933801e2dd3f8b605cab370

    [Overview of Choreo Connect]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/getting-started/choreo-connect-overview). 

## Via API Manager

### Step 1 - Define Virtual Hosts

Let's define virtual hosts (VHosts) in API Manager server instance by editing the `deployment.toml`.

!!! info
    Refer [Define Custom Hostnames]({{base_path}}/deploy-and-publish/deploy-on-gateway/deploy-api/exposing-apis-via-custom-hostnames/#step-1-define-the-custom-hostnames)
    for more information.

1. Open `<APIM-HOME>/repository/conf/deployment.toml` file.
2. Add the following config under the **Default** `[[apim.gateway.environment]]` to define the VHost `us.wso2.com`.
    ```toml
    [[apim.gateway.environment.virtual_host]]
    ws_endpoint = "ws://us.wso2.com:9099"
    wss_endpoint = "wss://us.wso2.com:8099"
    http_endpoint = "http://us.wso2.com/gateway"
    https_endpoint = "https://us.wso2.com/gateway"
    websub_event_receiver_http_endpoint = "http://us.wso2.com:9021"
    websub_event_receiver_https_endpoint = "https://us.wso2.com:8021"
    ``` 
### Step 2 - Configure Choreo Connect with API Manager

> Please refer [here]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/concepts/apim-as-control-plane/#configure-choreo-connect-with-api-manager).

Reviewer: ```suggestion
Refer to [documentation on how to configure Choreo Connect with API Manager]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/concepts/apim-as-control-plane/#configure-choreo-connect-with-api-manager).
```

Author: fixed with b9900b3

### Step 3 - Create an API in API Manager

Follow the steps [here]({{base_path}}/design/create-api/create-rest-api/create-a-rest-api-from-an-openapi-definition/).

### Step 4 - Deploy the API in API Manager

> The guide [here]({{base_path}}/deploy-and-publish/deploy-on-gateway/deploy-api/deploy-an-api) will explain how you could easily deploy the API you just created.

Reviewer: ```suggestion
The guide [here]({{base_path}}/deploy-and-publish/deploy-on-gateway/deploy-api/deploy-an-api) will explain how you can easily deploy the API you just created.
```

Author: fixed with b9900b3

> When deploying the API select the Virtual Host you defined earlier (i.e. `us.wso2.com`).

Reviewer: ```suggestion
When deploying the API, select the Virtual Host you defined earlier (i.e. `us.wso2.com`).
```

Author: fixed with b9900b3

> You have successfully deployed the API to the choreo connect with the VHost `us.wso2.com`!

Reviewer: ```suggestion
You have successfully deployed the API to Choreo Connect with the VHost `us.wso2.com`.
```

Author: fixed with b9900b3

> To invoke the API jump to the steps [here](#invoke-the-api).

Reviewer: ```suggestion
To invoke the API, skip to the steps [here](#invoke-the-api).
```

Author: fixed with b9900b3

## Via the API Controller (apictl) 

> Follow the all steps except **Deploy API** in the (Deploy API to Choreo Connect with apictl)[{{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/deploy-api/deploy-api-to-choreo-connect/#via-the-cli-tool-apictl].

Reviewer: ```suggestion
Follow the all steps except **Deploy API** in the (Deploy API to Choreo Connect with apictl documentation)[{{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/deploy-api/deploy-api-to-choreo-connect/#via-the-cli-tool-apictl].
```

Author: fixed with b9900b3

> Before deploying the API project lets edit the file `deployment_environments.yaml`.

Reviewer: ```suggestion
Before deploying the API project, edit the file `deployment_environments.yaml`.
```

Author: fixed with b9900b3

```yaml
type: deployment_environments
version: v4.0.0
data:
 -
   displayOnDevportal: true
   deploymentEnvironment: Default
   deploymentVhost: us.wso2.com
```

!!! info

>     When configuring multiple Choreo Connect gateway environments, you have to configure the default VHost of the particular environment.

Reviewer: ```suggestion
    When configuring multiple Choreo Connect Gateway environments, you have to configure the default VHost of the particular environment.
```

Author: fixed with b9900b3

    
    ```toml tab="Format"
    # default vhosts mapping for standalone mode
    [[adapter.vhostMapping]]
      environment = <ENVIRONMENT_NAME>
      vhost = <DEFAULT_VHOST_OF_ENVIRONMENT>
    ```
    
    ```toml tab="Example"
    # default vhosts mapping for standalone mode
    [[adapter.vhostMapping]]
      environment = "Default"
      vhost = "localhost"
    [[adapter.vhostMapping]]
      environment = "sg-region"
      vhost = "sg.wso2.com"
    ```
    

>     If the vhost is not declared in the API project, the API is deployed with the default vhost of the environment.

Reviewer: ```suggestion
    If the VHost is not declared in the API project, the API is deployed with the default VHost of the environment.
```

Author: fixed with b9900b3

>     For exammple, an API project with following `deployment_environments.yaml` will be deployed to following.

Reviewer: ```suggestion
    For exammple, an API project with following `deployment_environments.yaml` will be deployed to the following.
```

Author: fixed with b9900b3
